File: exercise/127/solution.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_map<string, int> reached;
        for (int i = 0; i < wordList.size(); i++) {
            reached[wordList[i]] = 0;
        }

        queue<pair<string, int>> fifo;
        fifo.push(make_pair(beginWord, 1));
        reached[beginWord] = 1;

        while (!fifo.empty()) {
            pair<string, int> now = fifo.front();
            fifo.pop();
            int length = now.second;
            for (int i = 0; i < now.first.length(); i++) {
                string word = now.first;
                for (char c = 'a'; c <= 'z'; c++) {
                    word[i] = c;
                    if (word == endWord) {
                        return length + 1;
                    }
                    if (reached.find(word) != reached.end() && reached[word] == 0) {
                        reached[word] = 1;
                        fifo.push(make_pair(word, length + 1));
                    }
                }
            }
        }
        return 0;
    }
};

File: exercise/127/solution.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        wordList.remove(endWord)
        wordSet = set(wordList)
        now_queue = [beginWord]
        back_queue = [endWord]
        layer = 2
        while len(now_queue) > 0:
            next_queue = []
            for word in now_queue:
                for i in range(len(word)):
                    new_word_list = list(word)
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        new_word_list[i] = c
                        new_word = ''.join(new_word_list)
                        if new_word in back_queue:
                            return layer
                        if new_word in wordSet:
                            next_queue.append(new_word)
                            wordSet.remove(new_word)
            if len(next_queue) > len(back_queue):
                now_queue = back_queue
            else:
                now_queue = next_queue
            layer += 1
        return 0
